ISSUE-774: Make reader and writer schema version mandatory for kafka

priyank5485 opened the pull request:
Change-Id: Ieb8b2195f51018ed97a39993c8719da07faac982

priyank5485 commented:
@shahsank3t @satishd @harshach Please review when you get a chance.
@shahsank3t We will need changes on UI side as well corresponding to the changes in the component spec json file for kafka source and sink.

shahsank3t commented on bootstrap/components/sinks/kafka-sink-topology-component.json:
```diff
@@ -72,10 +72,11 @@
       {
```
We might want to keep this `writerSchemaVersion` below `topic` field to make a better user experience.

priyank5485 replied:
@shahsank3t Will do

satishd commented on streams/service/src/main/java/com/hortonworks/streamline/streams/service/SchemaResource.java:
```diff
@@ -132,4 +133,61 @@ public Response getKafkaSourceSchema(@PathParam("topicName") String topicName,
+                    (version)));
+            if (schemaVersionInfo != null) {
+                LOG.debug("######### Received schema version from schema registry: {}", schemaVersionInfo);
+                String schema = schemaVersionInfo != null ? schemaVersionInfo.getSchemaText() : null;
```
no need to have this check as it is already done in the enclosing if condition.

priyank5485 replied:
Will fix it

satishd commented on streams/service/src/main/java/com/hortonworks/streamline/streams/service/SchemaResource.java:
```diff
@@ -132,4 +133,61 @@ public Response getKafkaSourceSchema(@PathParam("topicName") String topicName,
+            }
+        } catch (SchemaNotFoundException e) {
+            LOG.error("Schema not found: [{}]", schemaName, e);
+            throw EntityNotFoundException.byId(schemaName);
```
This should be `EntityNotFoundException#byName` instead of `EntityNotFoundException#byId`

priyank5485 replied:
Will change

satishd commented on streams/service/src/main/java/com/hortonworks/streamline/streams/service/SchemaResource.java:
```diff
@@ -132,4 +133,61 @@ public Response getKafkaSourceSchema(@PathParam("topicName") String topicName,
+        try {
+            LOG.info("Get all versions for schema : {}", schemaName);
+            Collection<SchemaVersionInfo> schemaVersionInfos = schemaRegistryClient.getAllVersions(schemaName);
+            Collection<String> schemaVersions = new ArrayList<>();
```
Is there any reason for using `Collection<String>` instead of `Collection<Integer>` ?

priyank5485 replied:
@satishd Because of the flux translation layer its easiest to make everything a string. Thats the approach I have taken for all kafka consumer and producer related properties - to keep them as string and handle the conversion where needed.

satishd commented on streams/service/src/main/java/com/hortonworks/streamline/streams/service/SchemaResource.java:
```diff
@@ -132,4 +133,61 @@ public Response getKafkaSourceSchema(@PathParam("topicName") String topicName,
+                LOG.debug("######### Converted schema: {}", schema);
+                return WSUtils.respondEntity(schema, OK);
+            } else {
+                return WSUtils.respondEntity("", NOT_FOUND);
```
may want to throw `SchemaNotFoundException ` here so that it throws right error to the client.

priyank5485 replied:
Will change

satishd commented on bootstrap/components/sinks/kafka-sink-topology-component.json:
```diff
@@ -72,10 +72,11 @@
         "uiName": "Writer schema version",
         "fieldName": "writerSchemaVersion",
-        "isOptional": true,
+        "isOptional": false,
```
Do not we have any issue while importing earlier streamline topologies as they may not have this configuration?

satishd commented on bootstrap/components/sources/kafka-source-topology-component.json:
```diff
@@ -37,10 +37,11 @@
         "uiName": "Reader schema version",
         "fieldName": "readerSchemaVersion",
-        "isOptional": true,
+        "isOptional": false,
```
same comment as earlier about importing earlier version topologies.

priyank5485 replied:
@satishd I dont think so. We are not exporting any component spec jsons. We should be okay since the fieldName has not changed.

satishd replied:
@priyank5485 I understand component spec jsons are not exported but the fields mentioned in uispec are represented by respective components which are stored in the backend storage. Earlier topology config's  kafka source will not have `readerSchemaVersion` in their config and when UI tries to load that in editor it may recognize that a mandatory field is missing and it may be throwing validation error. 

@shahsank3t @priyank5485 Pl let me know if I am missing anything here. I am just cautious whether we are breaking any topologies saved in earlier versions before this PR change.

shahsank3t replied:
@satishd 
Two scenarios:
1. If user imports the topology and directly tries to deploy, that will still work since all the streams/edges and everything was properly linked (unless on runtime backend does a validation on every component field whether a mandatory one is missing or not)

2. If user imports and opens the component (in this case Kafka source), the missing field will be blank and if the user doesn't add value for that field, UI won't allow to save & keeps showing `Required Field` error.

priyank5485 replied:
@satishd Never hurts to be cautious. Glad we discussed it. Like @shahsank3t confirmed, this wont be an issue. Merging this in

satishd replied:
@priyank5485 @shahsank3t I do not think current experience is seamless, it is still kind of broken as user needs to update with reader schema version when topology is edited. 
Ideally, user should not get any validation error when importing a topology created in earlier streamline versions. It should be upgraded seamlessly by showing respective info messages. In this case, this should have been upgraded by setting the right reader version(fetch from SR by sending the current schema text and its schema name) without asking user to configure it. 
This is not really a UI issue but a general import issue which should be handled by respective sources/sinks/processors, filed issue #822 on the same.

priyank5485 replied:
@satishd This change is kind of a backwards incompatible so its not possible to have a seamless experience. Moreover, I think UI forcing the user to pick a reader schema version in newer version is not a big inconvenience. I agree with you on showing an info message. However, setting the right reader version approach you suggested is not clear and seems like an overkill.  What is the right reader version here? You mentioned you get that by sending current schema text and name and get it from SR. Where do we get schema text from? And which version? By default in earlier versions if reader schema version is not specified the writer schema version from payload is used to deserialize.

priyank5485 commented:
@satishd @shahsank3t I have addressed your review comments. Please review again. @shahsank3t I have included your UI patch as a commit in this PR, however webservice module is failing compilation due to a javascript error in Fields.jsx Can you fix it?

priyank5485 commented:
@shahsank3t Fixed the UI build issue

shahsank3t commented:
+1
